File: python_backend/services/training_data_service.py
"""
Training Data Service - Historical data management and analytics
"""
import asyncio
from datetime import datetime, timedelta
from typing import Dict, List, Any, Optional
import sys
import os

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from database import async_session_maker
from models import SensorReading
from sqlalchemy import select, func, extract
from sqlalchemy.ext.asyncio import AsyncSession


class TrainingDataService:
    """Service for managing historical training data and providing analytics"""
    
    def __init__(self):
        pass
    
    async def get_training_data(
        self,
        start_date: datetime,
        end_date: datetime
    ) -> List[SensorReading]:
        """Get historical readings for ML model training"""
        async with async_session_maker() as session:
            result = await session.execute(
                select(SensorReading)
                .where(SensorReading.timestamp >= start_date)
                .where(SensorReading.timestamp <= end_date)
                .order_by(SensorReading.timestamp)
            )
            return list(result.scalars().all())
    
    async def get_last_n_days_of_data(self, days: int) -> List[SensorReading]:
        """Get last N days of data for training"""
        end_date = datetime.now()
        start_date = end_date - timedelta(days=days)
        return await self.get_training_data(start_date, end_date)
    
    async def get_average_power_by_hour(self) -> Dict[int, float]:
        """Get average power consumption by hour of day"""
        async with async_session_maker() as session:
            result = await session.execute(
                select(
                    extract('hour', SensorReading.timestamp).label('hour'),
                    func.avg(SensorReading.current_wattage).label('avg_wattage')
                )
                .group_by(extract('hour', SensorReading.timestamp))
            )
            rows = result.all()
            return {int(row.hour): row.avg_wattage for row in rows}
    
    async def get_power_statistics(
        self,
        start_date: datetime,
        end_date: datetime
    ) -> Dict[str, Any]:
        """Get power statistics for a specific time range"""
        readings = await self.get_training_data(start_date, end_date)
        
        if not readings:
            return {}
        
        wattages = [r.current_wattage for r in readings]
        
        stats = {
            "count": len(readings),
            "min": min(wattages),
            "max": max(wattages),
            "avg": sum(wattages) / len(wattages),
            "total_kwh": sum(wattages) / 1000,
        }
        
        # Calculate efficiency stats
        efficiency_scores = [r.efficiency_score for r in readings if r.efficiency_score is not None]
        if efficiency_scores:
            stats["avg_efficiency"] = sum(efficiency_scores) / len(efficiency_scores)
        
        # Count anomalies
        anomaly_count = sum(1 for r in readings if r.integrity_alert)
        stats["anomaly_count"] = anomaly_count
        stats["anomaly_rate"] = (anomaly_count / len(readings)) * 100 if readings else 0
        
        return stats
    
    async def get_typical_occupancy_by_hour(self) -> Dict[int, int]:
        """Get typical occupancy patterns by hour"""
        readings = await self.get_last_n_days_of_data(7)
        
        occupancy_by_hour: Dict[int, List[int]] = {i: [] for i in range(24)}
        
        for reading in readings:
            hour = reading.timestamp.hour
            occupancy_by_hour[hour].append(reading.occupancy_count)
        
        typical_occupancy = {}
        for hour, values in occupancy_by_hour.items():
            if values:
                values.sort()
                median = values[len(values) // 2]
                typical_occupancy[hour] = median
        
        return typical_occupancy
    
    async def get_peak_usage_hours(self) -> List[int]:
        """Get peak usage hours"""
        hourly_averages = await self.get_average_power_by_hour()
        
        sorted_hours = sorted(hourly_averages.items(), key=lambda x: x[1], reverse=True)
        return [hour for hour, _ in sorted_hours[:5]]
    
    async def get_off_peak_usage_hours(self) -> List[int]:
        """Get off-peak usage hours (lowest consumption)"""
        hourly_averages = await self.get_average_power_by_hour()
        
        sorted_hours = sorted(hourly_averages.items(), key=lambda x: x[1])
        return [hour for hour, _ in sorted_hours[:5]]
    
    async def get_daily_consumption_summary(self, days: int) -> List[Dict[str, Any]]:
        """Get daily energy consumption summary"""
        daily_summaries = []
        now = datetime.now()
        
        for i in range(days - 1, -1, -1):
            day_start = (now - timedelta(days=i)).replace(
                hour=0, minute=0, second=0, microsecond=0
            )
            day_end = day_start + timedelta(days=1)
            
            stats = await self.get_power_statistics(day_start, day_end)
            if stats:
                stats["date"] = day_start.strftime("%Y-%m-%d")
                daily_summaries.append(stats)
        
        return daily_summaries
    
    async def get_data_quality_metrics(self) -> Dict[str, Any]:
        """Get data quality metrics"""
        async with async_session_maker() as session:
            result = await session.execute(select(SensorReading))
            all_readings = list(result.scalars().all())
        
        total_readings = len(all_readings)
        metrics = {"total_readings": total_readings}
        
        if total_readings == 0:
            metrics["quality_score"] = 0
            return metrics
        
        # Check for missing fields
        readings_with_temp = sum(1 for r in all_readings if r.external_temp is not None)
        readings_with_efficiency = sum(1 for r in all_readings if r.efficiency_score is not None)
        
        metrics["readings_with_temperature"] = readings_with_temp
        metrics["temperature_coverage_pct"] = (readings_with_temp / total_readings) * 100
        
        metrics["readings_with_efficiency"] = readings_with_efficiency
        metrics["efficiency_coverage_pct"] = (readings_with_efficiency / total_readings) * 100
        
        # Overall quality score (0-100)
        quality_score = (
            (readings_with_temp / total_readings) * 50 +
            (readings_with_efficiency / total_readings) * 50
        )
        metrics["quality_score"] = round(quality_score, 2)
        
        return metrics

File: ai_engine/backend/backend/src/main/java/com/ecosync/service/TrainingDataService.java
package com.ecosync.service;

import com.ecosync.model.SensorReading;
import com.ecosync.repository.SensorReadingRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.util.*;
import java.util.stream.Collectors;

/**
 * Service for managing historical training data and providing analytics
 */
@Service
public class TrainingDataService {

    @Autowired
    private SensorReadingRepository sensorReadingRepository;

    /**
     * Get historical readings for ML model training
     */
    public List<SensorReading> getTrainingData(LocalDateTime startDate, LocalDateTime endDate) {
        return sensorReadingRepository.findReadingsForTraining(startDate, endDate);
    }

    /**
     * Get last N days of data for training
     */
    public List<SensorReading> getLastNDaysOfData(int days) {
        LocalDateTime endDate = LocalDateTime.now();
        LocalDateTime startDate = endDate.minusDays(days);
        return getTrainingData(startDate, endDate);
    }

    /**
     * Get average power consumption by hour of day
     */
    public Map<Integer, Double> getAveragePowerByHour() {
        List<Object[]> results = sensorReadingRepository.getAverageWattageByHour();
        Map<Integer, Double> hourlyAverages = new HashMap<>();
        for (Object[] result : results) {
            Integer hour = ((Number) result[0]).intValue();
            Double avgWattage = (Double) result[1];
            hourlyAverages.put(hour, avgWattage);
        }
        return hourlyAverages;
    }

    /**
     * Get power statistics for a specific time range
     */
    public Map<String, Object> getPowerStatistics(LocalDateTime start, LocalDateTime end) {
        List<SensorReading> readings = getTrainingData(start, end);
        
        if (readings.isEmpty()) {
            return Collections.emptyMap();
        }

        List<Double> wattages = readings.stream()
                .map(SensorReading::getCurrentWattage)
                .collect(Collectors.toList());

        Map<String, Object> stats = new HashMap<>();
        stats.put("count", readings.size());
        stats.put("min", Collections.min(wattages));
        stats.put("max", Collections.max(wattages));
        stats.put("avg", wattages.stream().mapToDouble(Double::doubleValue).average().orElse(0));
        stats.put("total_kwh", wattages.stream().mapToDouble(Double::doubleValue).sum() / 1000);
        
        // Calculate efficiency stats
        List<Double> efficiencyScores = readings.stream()
                .filter(r -> r.getEfficiencyScore() != null)
                .map(SensorReading::getEfficiencyScore)
                .collect(Collectors.toList());
        
        if (!efficiencyScores.isEmpty()) {
            stats.put("avg_efficiency", efficiencyScores.stream()
                    .mapToDouble(Double::doubleValue).average().orElse(0));
        }

        // Count anomalies
        long anomalyCount = readings.stream()
                .filter(r -> Boolean.TRUE.equals(r.getIntegrityAlert()))
                .count();
        stats.put("anomaly_count", anomalyCount);
        stats.put("anomaly_rate", (double) anomalyCount / readings.size() * 100);

        return stats;
    }

    /**
     * Get typical occupancy patterns by hour
     */
    public Map<Integer, Integer> getTypicalOccupancyByHour() {
        List<SensorReading> lastWeek = getLastNDaysOfData(7);
        
        Map<Integer, List<Integer>> occupancyByHour = new HashMap<>();
        for (int i = 0; i < 24; i++) {
            occupancyByHour.put(i, new ArrayList<>());
        }
        
        for (SensorReading reading : lastWeek) {
            int hour = reading.getTimestamp().getHour();
            occupancyByHour.get(hour).add(reading.getOccupancyCount());
        }
        
        Map<Integer, Integer> typicalOccupancy = new HashMap<>();
        for (Map.Entry<Integer, List<Integer>> entry : occupancyByHour.entrySet()) {
            List<Integer> values = entry.getValue();
            if (!values.isEmpty()) {
                // Use median as typical value
                Collections.sort(values);
                int median = values.get(values.size() / 2);
                typicalOccupancy.put(entry.getKey(), median);
            }
        }
        
        return typicalOccupancy;
    }

    /**
     * Get peak usage hours
     */
    public List<Integer> getPeakUsageHours() {
        Map<Integer, Double> hourlyAverages = getAveragePowerByHour();
        
        return hourlyAverages.entrySet().stream()
                .sorted(Map.Entry.<Integer, Double>comparingByValue().reversed())
                .limit(5)
                .map(Map.Entry::getKey)
                .collect(Collectors.toList());
    }

    /**
     * Get off-peak usage hours (lowest consumption)
     */
    public List<Integer> getOffPeakUsageHours() {
        Map<Integer, Double> hourlyAverages = getAveragePowerByHour();
        
        return hourlyAverages.entrySet().stream()
                .sorted(Map.Entry.comparingByValue())
                .limit(5)
                .map(Map.Entry::getKey)
                .collect(Collectors.toList());
    }

    /**
     * Get daily energy consumption summary
     */
    public List<Map<String, Object>> getDailyConsumptionSummary(int days) {
        List<Map<String, Object>> dailySummaries = new ArrayList<>();
        LocalDateTime now = LocalDateTime.now();
        
        for (int i = days - 1; i >= 0; i--) {
            LocalDateTime dayStart = now.minusDays(i).toLocalDate().atStartOfDay();
            LocalDateTime dayEnd = dayStart.plusDays(1);
            
            Map<String, Object> stats = getPowerStatistics(dayStart, dayEnd);
            if (!stats.isEmpty()) {
                stats.put("date", dayStart.toLocalDate().toString());
                dailySummaries.add(stats);
            }
        }
        
        return dailySummaries;
    }

    /**
     * Clean up old readings (keep only last N days)
     */
    public int cleanupOldReadings(int keepDays) {
        LocalDateTime cutoff = LocalDateTime.now().minusDays(keepDays);
        List<SensorReading> allReadings = sensorReadingRepository.findAll();
        
        int deleted = 0;
        for (SensorReading reading : allReadings) {
            if (reading.getTimestamp().isBefore(cutoff)) {
                sensorReadingRepository.delete(reading);
                deleted++;
            }
        }
        
        return deleted;
    }

    /**
     * Get data quality metrics
     */
    public Map<String, Object> getDataQualityMetrics() {
        Map<String, Object> metrics = new HashMap<>();
        
        long totalReadings = sensorReadingRepository.count();
        metrics.put("total_readings", totalReadings);
        
        if (totalReadings == 0) {
            metrics.put("quality_score", 0);
            return metrics;
        }
        
        // Check for missing fields
        List<SensorReading> allReadings = sensorReadingRepository.findAll();
        
        long readingsWithTemp = allReadings.stream()
                .filter(r -> r.getExternalTemp() != null)
                .count();
        
        long readingsWithEfficiency = allReadings.stream()
                .filter(r -> r.getEfficiencyScore() != null)
                .count();
        
        metrics.put("readings_with_temperature", readingsWithTemp);
        metrics.put("temperature_coverage_pct", (double) readingsWithTemp / totalReadings * 100);
        
        metrics.put("readings_with_efficiency", readingsWithEfficiency);
        metrics.put("efficiency_coverage_pct", (double) readingsWithEfficiency / totalReadings * 100);
        
        // Overall quality score (0-100)
        double qualityScore = ((double) readingsWithTemp / totalReadings * 50) +
                             ((double) readingsWithEfficiency / totalReadings * 50);
        metrics.put("quality_score", Math.round(qualityScore * 100) / 100.0);
        
        return metrics;
    }
}
